rfd/0043-kubeaccess-multiparty.md

```diff
@@ -0,0 +1,386 @@
+
+To make this process easier for the user. I propose extending the current `tsh join` command
+to also work for Kubernetes access in the form of `tsh kube join <session-id>`. This attaches
+to an ongoing session and displays stdout/stderr.
+
+##### MFA tap
+
+If the standard `per_session_mfa` option is enabled for a role then MFA tap input via Yubikey or other is required for the participant to be considered active.
+This requirement is on an interval of 1 minute. When there is 15 seconds left, an alert is printed to the console.
+
+```
+Teleport >> Please tap your MFA key within 15 seconds.
+```
+
+If tap is made after the alert, the follwing message is shown:
+
+```
+Teleport >> MFA tap received.
+```
+
+##### Example
+
+This example illustrates how a group 3 users of which Alice is the initiator and Eve and Ben are two observers
+start a multiparty session. Below is a a series of events that happen that include what each user sees and what
+they do.
+
+- Alice initiates an interactive session to a pod: `tsh kube exec --invite=ben@foo.net,eve@foo.net,alice@foo.net -- database_pod -- /bin/bash`
+- Alice sees:
+```
+Creating session with uuid <example-uuid>...
+This session requires moderator. Waiting for others to join:
+- role: auditor-role x2
+```
+- Eve joins the session with `tsh kube join <example-uuid>` and sees:
+```
+Please tap MFA key to continue...
+```
+- Eve taps MFA
+- Alice and Eve sees:
+```
+Creating session with uuid <example-uuid>...
+This session requires moderator. Waiting for others to join:
+- role: auditor-role x1
+Events:
+- User Eve joined the session.
+```
+- Ben joins the session with `tsh kube join <example-uuid>` and sees:
+```
+Please tap MFA key to continue...
+```
+- Ben taps MFA
+- Alice, Eve and Ben sees
+```
+Creating session with uuid <example-uuid>...
+Session starting...
+Events:
+- User Eve joined the session.
+- User Ben joined the session
+```
+- The connection to the pod is made and each the session turns into a normal shell.
+
+#### Session invites and notifications
+
+Shared sessions for Kubernetes access will have support for participant invites and notifications.
+Ongoing sessions are tracked and can be listed to make it easier to find and join them.
+
+Ongoing sessions you have access to view can be listed with `tsh kube sessions ls`.
+This easily allows eligible participants to find and join a session waiting for participants easily
+
+When the `--invite` flag is used with `tsh kube exec`, the invitees are tracked and included in the
+session resource which allows Teleport clients and plugins to detect notify them.
+
+##### Session resource
+
+There currently isn't a general purpose session resource in Teleport that's suitable.
+therefore I suggest that this shall be added. This will be initially used for tracking Kubernetes
+sessions but is compatible with all current and future session types.
+
+```protobuf
+// SessionSpecV3 is the specification for a live session.
+message SessionSpecV3 {
```

Comment:
How does it correlate with existing session resource? Are we going to have two session resources with addition of this? Do we want to extend the existing one?

https://github.com/gravitational/teleport/blob/master/lib/session/session.go

Reply:
So my idea was that we take this current session resource system which isn't really a resource and we convert it to a real gRPC defined resource like others like shown. The job of this resource is to list active sessions and their metadata needed for `tsh kube sessions` to work and for integration with listing sessions in the Web UI.

We'd then store more detailed info like terminal params and net streams that's only needed to connect on the node/agent handling the multiplexing logic to avoid constantly updating the resource.

That said it may be easier to simply modify this without introducing a gRPC conversion but it seems that we're moving everything over to gRPC. We'd also define update messages in gRPC that are send by the multiplexing node/agent that are used to update the resource when participants and such change.

Reply:
I've gone ahead and clarified in the RFD on how I intend for this to to correlate with the existing resource and how information is tracked.